docs: add guide for developing Lua plugins (#29073)
neovim/neovim: runtime/doc/lua-plugin.txt

                                       Type |gO| to see the table of contents.

==============================================================================
Introduction                                                       *lua-plugin*

This is a guide for getting started with Nvim plugin development. It is not
intended as a set of rules, but as a collection of recommendations for good
practices.

For a guide to using Lua in Nvim, please refer to |lua-guide|.

==============================================================================
Type safety                                            *lua-plugin-type-safety*

Lua, as a dynamically typed language, is great for configuration. It provides
virtually immediate feedback.
But for larger projects, this can be a double-edged sword, leaving your plugin
susceptible to unexpected bugs at the wrong time.

You can leverage LuaCATS https://luals.github.io/wiki/annotations/
annotations, along with lua-language-server https://luals.github.io/ to catch
potential bugs in your CI before your plugin's users do.

------------------------------------------------------------------------------
Tools						 *lua-plugin-type-safety-tools*

> - lua-typecheck-action https://github.com/marketplace/actions/lua-typecheck-action

Reviewer: IMO We should provide recommendations for a terminal-only approach to typechecking as well as the **along with** the GitHub action so people know how to do both. We could provide a minimal repository example and link it here. The jist is

```json
{
    "diagnostics.libraryFiles": "Disable",
    "runtime.version": "LuaJIT",
    "workspace.library": [
        "$VIMRUNTIME/lua"
    ]
}
```

+ running `llscheck` is all users need, I've found. `nvim-best-practices-plugin-template` wraps some extra things on top ([Makefile](https://github.com/ColinKennedy/nvim-best-practices-plugin-template/blob/main/Makefile#L25C59-L25C68) + [.luarc.json](https://github.com/ColinKennedy/nvim-best-practices-plugin-template/blob/main/.luarc.json)) but the minimal setup is really low effort and can be run easily from the terminal

Author: Thinking about it, ideally this kind of detailed info should go in the `llscheck` and `lua-typecheck-action` documentation.

I'd [like to make lua-typecheck-action use llschec](https://github.com/mrcjkb/lua-typecheck-action/issues/74), or perhaps migrate it to [`rocks`](https://github.com/nvim-neorocks/rocks), once the `rocks lint` or `rocks typecheck` command has been implemented and `rocks` is ready for release (which should be soon).

Reviewer: If llscheck is receptive to have neovim-related docs in their repository then I suppose linking from here to there makes sense. The alternative though (neovim specific hello world repository) I think is still useful either way. The Neovim (core / community) should own its own resources.

      specific file types or floating windows.

A common approach to allow keymap configuration is to define a declarative DSL
https://en.wikipedia.org/wiki/Domain-specific_language via a `setup` function.

However, doing so means that

- You will have to implement and document it yourself.
- Users will likely face inconsistencies if another plugin has a slightly
  different DSL.
- |init.lua| scripts that call such a `setup` function may throw an error if
  the plugin is not installed or disabled.

As an alternative, you can provide |<Plug>| mappings to allow users to define
their own keymaps with |vim.keymap.set()|.

- This requires one line of code in user configs.
- Even if your plugin is not installed or disabled, creating the keymap won't
  throw an error.

Another option is to simply expose a Lua function or |user-commands|.

However, some benefits of |<Plug>| mappings over this are that you can

- Enforce options like `expr = true`.
- Expose functionality only for specific |map-modes|.
- Expose different behavior for different |map-modes| with a single |<Plug>|
  mapping, without adding impurity or complexity to the underlying Lua

>   implementation.

Reviewer 2: This may have been discussed at length already, but I am wondering whether the `<Plug>(...)` mapping should be officially recommended.

Today's plugins are becoming more and more sophisticated, and I think that a DSL-based definition method like the `<Plug>(...)` mapping is not flexible enough.

For example, `vim.diagnostic.jump()` is a good example of this discussion.

```lua
vim.keymap.set('n', '<C-j>', function()
  vim.diagnostic.jump({
    count = 1,
    float = true,
    wrap = false,
  })
end)
```

It is difficult to achieve this ergonomically with the `<Plug>(...)` mapping. (Functional methods also enjoy the benefits of static type checking and completion.)

I understand the opinion that "there are also advantages to `<Plug>(...)`", but I wonder if neovim is making a decision to recommend `<Plug>(...)`.

Reviewer: I think the idea of `<Plug>`  is to provide a fixed API whose implementation can change underneath without disturbing the user. You're right that people could use native Lua to do the same thing but in practice rarely do.

For example a plugin like

```
{root}/
    - lua/
        - my_plugin/
            - core/
                - movement.lua
            - init.lua
```

`{root}/lua/my_plugin/core/movement.lua`
```lua
local M = {}
function M.jump(options) ... end
return M
```

`{root}/lua/my_plugin/init.lua`
```lua
local M = {}

-- If this was a <Plug>, it'd be under the `plugin` folder and be called  '<Plug>(MyPlugin-Jump-Backwards-Once)'
function M.jump_backwards_once()
    require("my_plugin.core.movement").jump({forwards=true, count=1})
end

-- If this was a <Plug>, it'd be under the `plugin` folder and be called  '<Plug>(MyPlugin-Jump-Forwards-Once)'
function M.jump_forwards_once()
    require("my_plugin.core.movement").jump({forwards=true, count=1})
end

return M
```

Later in a user's configuration

```lua
vim.keymap.set('n', '<C-k>', function() require("my_plugin").jump_backwards_once() end)
vim.keymap.set('n', '<C-j>', function() require("my_plugin").jump_forwards_once() end)

--- Versus
vim.keymap.set('n', '<C-j>', '<Plug>(MyPlugin-Jump-Backwards-Once)')
vim.keymap.set('n', '<C-j>', '<Plug>(MyPlugin-Jump-Forwards-Once)')
```

The Lua function isn't really doing anything conceptually different from `<Plug>`. You could probably make a case that the Lua way is better because it's "more lazy-loaded".

Reviewer 2: I think there is a big difference between the two in that "static checks are possible with LuaLS" and "completion works".

Also, it is difficult to actually convert the API of `vim.diagnostic.jump` to `<Plug>(...)`.

You can set options such as wrap, float, and servarity. It is not realistic to enumerate all combinations with `<Plug>(...)`.

I thought it would be better to avoid making it an official recommendation just because it works with a simplified sample.

However, if neovim officially recommends `<Plug>(...)`, I will follow that :)

Author: See [this comment about the relevance of `<Plug>`](https://github.com/nvim-neorocks/nvim-best-practices/issues/5#issuecomment-2131367234).

To address your specific concern about `vim.diagnostic.jump`, and as I've mentioned in a previous discussion about plugin initialisation:

Plugin authors can choose to diverge if they deem it necessary, which is made clear right at the beginning of this guide:

>   it is not intended as a set of rules...

Furthermore, I have already added a paragraph that should address your concern: 

> Another option is to simply expose a Lua function or |user-commands|.

If it's not clear from that that exposing a Lua function isn't a bad practise, then perhaps that needs to be made more clear?

About static type checking: It's nice to have, but yagni in a user config.

Reviewer 2: What I felt after reading this document

- There is an option to use DSL to define keymaps in the setup function
    - It is better to use `<Plug>`
- You can provide `Lua function` as another option
    - It is better to use `<Plug>`

So, I felt that this sentence means "neovim officially recommends using `<Plug>`".

I'm sorry if that was not what I intended.

Author: Then it probably needs some rephrasing :smile: 
I'll look into that after work.

Reviewer 2: I'm going a bit off-topic, but I'd like to share my thoughts.

I still prefer exposing Lua functions over the <Plug>(...) approach.
Perhaps it's because I prioritize extensibility the most as a plugin developer.
In any case, I hope this won't be officially regarded as a "discouraged practice."

I strongly agree with not recommending a key mapping DSL in the setup function.
(Yes, I am the author of nvim-cmp. I implemented some hacks to support commitCharacters and something like user-defined modes, but I regret doing so.)

Author: I've rephrased the text to hopefully clarify that I'm recommending either `<Plug>` or a Lua function, or both, depending on the use case:

This is the updated text:

```
Another option is to simply expose a Lua function or |user-commands|.

Some benefits of |<Plug>| mappings are that you can

- Enforce options like `expr = true`.
- Expose functionality only for specific |map-modes|.
- Expose different behavior for different |map-modes| with a single |<Plug>|
  mapping, without adding impurity or complexity to the underlying Lua
  implementation.

Some benefits of exposing a Lua function are:

- Extensibility, if the function takes an options table as an argument.
- A cleaner UX, if there are many options and enumerating all combinations
  of options would result in a lot of |<Plug>| mappings.

NOTE: If your function takes an options table, users may still benefit 
      from |<Plug>| mappings for the most common combinations.
```

Reviewer 3: Note that you can achieve the same benefits with native Lua functions; you just have to separate your internal API from functions you (re-)export e.g. via wrappers (around `require`d modules) )defined in your `plugin`.

Reviewer 4: Unsure if this has been discussed before, but it's worth noting that `<Plug>` mappings have the benefit of being strings, which means they can be looked up using `vim.fn.hasmapto` (I don't believe you can do the same for function references, unless neovim does some internal magic to allow this?). This means plugins can check if a `<Plug>` is mapped and not remap their own standard keys, instead respecting the user's override.

Author: > Note that you can achieve the same benefits with native Lua functions;

It's possible, but it involves adding impure checks (like `vim.api.nvim_get_mode()`) to the function (or wrapper function) implementation, whereas the `vim.keymap` interface handles mode-specific behavior as part of its core API.
To my knowledge, there's no way to enforce `expr = true` with a Lua function? (I may be mistaken).

[`<Plug>` mappings aren't Vimscript, they're a mapping feature.](https://github.com/nvim-neorocks/nvim-best-practices/issues/5#issuecomment-2131367234)

> separating public and internal API to avoid leaking implementation details.

This is an important point, but you can do that with a DSL in a `setup` function too.
I wrote this section to be about Lua and/or `<Plug>` mappings vs. DSLs in a `setup` function.

I recommend `<Plug>` mappings _as an option_, because they are underutilised,
due to the common misconception that they're an "outdated Vimscript" thing.
Again, this is _not a recommendation against exposing a Lua function_.

I've updated the section again, to be more clear:

```
Some benefits of |<Plug>| mappings are that you can

- Enforce options like `expr = true`.
- Use |vim.keymap|'s built-in mode handling to expose functionality only for
  specific |map-modes|.
- Handle different |map-modes| differently with a single mapping, without
  adding mode checks to the underlying implementation.
```

  default configuration and does not contain any initialization logic.
- A Vimscript compatible table (e.g. in the |vim.g| or |vim.b| namespace) that your
  plugin reads from and validates at initialization time.
  See also |lua-vim-variables|.

Typically, automatic initialization logic is done in a |plugin| or |ftplugin|
script. See also |'runtimepath'|.

==============================================================================
Lazy loading					      *lua-plugin-lazy-loading*

When it comes to initializing your plugin, assume your users may not be using
a plugin manager that takes care of lazy loading for you.
Making sure your plugin does not unnecessarily impact startup time is your
responsibility. A plugin's functionality may evolve over time, potentially
leading to breakage if users have to hack into the loading mechanisms.
Furthermore, a plugin that implements its own lazy initialization properly will
likely have less overhead than the mechanisms used by a plugin manager or user
to load that plugin lazily.

------------------------------------------------------------------------------
Defer `require` calls			*lua-plugin-lazy-loading-defer-require*

> |plugin| scripts should not eagerly `require` Lua modules.

Reviewer 5: I think this should also apply to the Lua module (assuming that it is the main module) containing `setup()` or `config()` function. That module should also be small (maybe only containing public functions)

And maybe this section should also tell users how to check startuptime of a plugin (using `nvim --startuptime` and look for modules of the plugin and its dependencies?)

Author: I've added a paragraph that says the same applies if a plugin has a Lua module as an entrypoint.
I don't think a detailed tutorial on `:h --startuptime` and `:h profile` is within the scope of this, but I've added a short note to the section:

> NOTE: You can use |--startuptime| to |profile| the impact a plugin has on
>       startup time.

their own keymaps with |vim.keymap.set()|.

- This requires one line of code in user configs.
- Even if your plugin is not installed or disabled, creating the keymap won't
  throw an error.

Another option is to simply expose a Lua function or |user-commands|.

Some benefits of |<Plug>| mappings are that you can

- Enforce options like `expr = true`.
- Use |vim.keymap|'s built-in mode handling to expose functionality only for
  specific |map-modes|.
- Handle different |map-modes| differently with a single mapping, without
  adding mode checks to the underlying implementation.
- Detect user-defined mappings through |hasmapto()| before creating defaults.

Some benefits of exposing a Lua function are:

- Extensibility, if the function takes an options table as an argument.
- A cleaner UX, if there are many options and enumerating all combinations
  of options would result in a lot of |<Plug>| mappings.

NOTE: If your function takes an options table, users may still benefit
      from |<Plug>| mappings for the most common combinations.

KEYMAP EXAMPLE

In your plugin:
>lua
    vim.keymap.set('n', '<Plug>(SayHello)', function()
        print('Hello from normal mode')

>     end, { noremap = true })

Reviewer 6: `noremap` in `vim.keymap.set` has no effect.

Maybe worth removing these in #35462 too?

  check, to reduce overhead.

==============================================================================
Troubleshooting                                   *lua-plugin-troubleshooting*

While developing a plugin, you can use the |:restart| command to see the
result of code changes in your plugin.

HEALTH

Nvim's "health" framework gives plugins a simple way to report status checks
to users. See |health-dev| for an example.

Basically, this just means your plugin will have a `lua/{plugin}/health.lua`
file. |:checkhealth| will automatically find this file when it runs.

Some things to validate:

- User configuration
- Proper initialization
- Presence of Lua dependencies (e.g. other plugins)
- Presence of external dependencies

MINIMAL CONFIG TEMPLATE

It can be useful to provide a template for a minimal configuration, along with
a guide on how to use it to reproduce issues.

==============================================================================
Versioning and releases                                *lua-plugin-versioning*

Consider:

- Use |vim.deprecate()| or a `---@deprecate` annotation when you need to

>   communicate a (future) breaking change or discourged practice.

Author: typo: discouraged (couldn't find this in #35462)

```suggestion
  communicate a (future) breaking change or discouraged practice.
```

        local foo = require("foo")
        foo.do_something()
    end, {
      -- ...
    })
<
NOTE: For a Vimscript alternative to `require`, see |autoload|.

NOTE: In case you are worried about eagerly creating user commands, autocommands
      or keymaps at startup:
      Plugin managers that provide abstractions for lazy-loading plugins on
      such events will need to create these themselves.

------------------------------------------------------------------------------
Filetype-specific functionality		     *lua-plugin-lazy-loading-filetype*

Consider making use of |filetype| for any functionality that is specific to a
filetype, by putting the initialization logic in a `ftplugin/{filetype}.lua`
script.

------------------------------------------------------------------------------
Example				     *lua-plugin-lazy-loading-filetype-example*

A plugin tailored to Rust development might have initialization in 
`ftplugin/rust.lua`:
>lua
    if not vim.g.loaded_my_rust_plugin then
        -- Initialize
    end
    -- NOTE: Using `vim.g.loaded_` prevents the plugin from initializing twice
    -- and allows users to prevent plugins from loading
    -- (in both Lua and Vimscript).

>     vim.g.loaded_my_rust_plugin = true

Reviewer 7: that's for thing which should just be done once globally, not per buffer

e.g. setting up autocommands, commands, mappings, config, ...

                                       Type |gO| to see the table of contents.

==============================================================================
Introduction                                                       *lua-plugin*

This is a guide for getting started with Nvim plugin development. It is not
intended as a set of rules, but as a collection of recommendations for good
practices.

For a guide to using Lua in Nvim, please refer to |lua-guide|.

==============================================================================
Type safety                                            *lua-plugin-type-safety*

Lua, as a dynamically typed language, is great for configuration. It provides
virtually immediate feedback.
But for larger projects, this can be a double-edged sword, leaving your plugin
susceptible to unexpected bugs at the wrong time.

You can leverage LuaCATS https://luals.github.io/wiki/annotations/
annotations, along with lua-language-server https://luals.github.io/ to catch
potential bugs in your CI before your plugin's users do.

------------------------------------------------------------------------------
Tools						 *lua-plugin-type-safety-tools*

> - lua-typecheck-action https://github.com/marketplace/actions/lua-typecheck-action

Reviewer 8: This also seems completely broken. Let's take it easy with the recommendations unless we're certain they're reliable.

Author: I maintain this one. How are you testing it?
It has a workflow that runs on itself: https://github.com/mrcjkb/lua-typecheck-action/actions/workflows/typecheck.yml
It just needs a properly configured .luarc.json for packages that have dependencies.

Reviewer 7: I use this and also know a few projects which use it where it works just fine

Reviewer 8: Oh, it defaults to checking `lua` directory which doesn't exist in neovim. Is there a reason we can't just check the actual repo root?

Reviewer 8: OK, you know what? I'll just open 10 issues in your repo and we can work on it there.

Reviewer 8: Right. I assume `llscheck` giving wonky results might also be due to the incorrect `.luarc.json` file?

Reviewer 8: @mrcjkb wanna send a separate PR to neovim to fix `.luarc.json` and `test/.luarc.json`? You seem to already have a good grasp of how these should be configured. (I looked at the docs for luals and got scared)

Author: Sure. I'm not quite sure how to fix the `${3rd}` libraries here. That would require [using addons](https://github.com/LuaLS/LLS-Addons). Maybe we should open an issue first?

Reviewer 8: https://github.com/neovim/neovim/pull/29115

> - llscheck https://github.com/jeffzi/llscheck

Reviewer 8: This seems absolutely broken. Have you tried this and gotten it to work? It gives completely different results if you run it for a directory `llscheck scripts` vs `llscheck scripts/*` and both of the are wrong, the `--configpath` parameter seems to do nothing. It's impressively bad. What is this doing here?

Author: This one hasn't been touched in a while, so yeah it's very possibly broken.

Reviewer 8: Perhaps my weird results with llscheck was due to our .luarc.json being incorrectly configured. Can keep in mind for later.

> - luacheck https://github.com/mpeterv/luacheck for additional linting

Reviewer 8: Repo is abandoned, link to maintained fork instead.

```suggestion
- luacheck https://github.com/lunarmodules/luacheck for additional linting
```

eagerly, you can lazy load it by moving the `require` into the command's
implementation.
>lua
    vim.api.nvim_create_user_command("MyCommand", function()
        local foo = require("foo")
        foo.do_something()
    end, {
      -- ...
    })
<
NOTE: For a Vimscript alternative to `require`, see |autoload|.

NOTE: In case you are worried about eagerly creating user commands, autocommands
      or keymaps at startup:
      Plugin managers that provide abstractions for lazy-loading plugins on
      such events will need to create these themselves.

------------------------------------------------------------------------------
Filetype-specific functionality		     *lua-plugin-lazy-loading-filetype*

Consider making use of |filetype| for any functionality that is specific to a
filetype, by putting the initialization logic in a `ftplugin/{filetype}.lua`
script.

------------------------------------------------------------------------------
Example				     *lua-plugin-lazy-loading-filetype-example*

A plugin tailored to Rust development might have initialization in 
`ftplugin/rust.lua`:
>lua
    if not vim.g.loaded_my_rust_plugin then

>         -- Initialise

Reviewer 8: ```suggestion
        -- Initialize
```

    end
    -- NOTE: Using `vim.g.loaded_` prevents the plugin from initializing twice
    -- and allows users to prevent plugins from loading
    -- (in both Lua and Vimscript).
    vim.g.loaded_my_rust_plugin = true

    local bufnr = vim.api.nvim_get_current_buf()
    -- do something specific to this buffer,
    -- e.g. add a |<Plug>| mapping or create a command
    vim.keymap.set("n", "<Plug>(MyPluginBufferAction)", function() 
        print("Hello")
    end, { noremap = true, buffer = bufnr, })
<
==============================================================================

> Configuration				             *lua-plugin-configuration*

Reviewer 8: This whole section feels out of place here. If there's some part of the validation functionality that is missing from `vim.validate`, then we should try to improve the functionality in core rather than have every plugin author implement/augment validation functionality on their own.

*lua-plugin.txt*                     Nvim

                            NVIM REFERENCE MANUAL

		  Guide to developing Lua plugins for Nvim

                                       Type |gO| to see the table of contents.

==============================================================================
Introduction                                                       *lua-plugin*

> This is a guide for getting started with Nvim plugin development. It is not

Reviewer 9: Reading this I got the impression that this is more an intermediate best practice guide and not a beginner getting started guide.  
(E.g compare with the `:h write-plugin` intro)

Author: How about  I remove the "getting started with"?

Reviewer 8: Let's change the content to be beginner friendly instead with a way to "advance" to the more intermediate topics.

Reviewer 8: I'll rereview it when I have more time with more useful comments.

Reviewer 9: I'm not sure if just removing "getting started" will cut it.

Overall this feels more like `plugin-ui-patterns` than `lua-plugin`.

It talks about:

- Best practices in regards to keymap definitions for users
    - Should tie into `lua-guide-mappings`
- Best practices on how to expose configuration options to users
- Best practices on how to design command APIs
    - Should tie into `lua-guide-commands`
    
- Performance considerations
    - Partly covered by `lua-guide-modules` and `lua-module-load` and also related to `autoload`

None of this is beginner content for how to get a plugin working.
And many parts are not exclusive to Lua.

What I'd expect from a "How to write a plugin" targeted towards beginner is:

- Language choices and link to resources to learn them (vim, lua, but also external providers)
- Entry points (runtime path, packages, plugin/, ftplugin, ftdetect, autocmd hooks)
- Available APIs

But a lot of this is covered via `lua-guide` and the various vim resources. I also suspect that writing something like this is not what Justin had in mind, but rather something that's closer to the style of `:h dev-patterns`

But let's wait for @justinmk to clarify what his intentions were.

intended as a set of rules, but as a collection of recommendations for good
practices.

For a guide to using Lua in Neovim, please refer to |lua-guide|.

==============================================================================
Type safety                                            *lua-plugin-type-safety*

Lua, as a dynamically typed language, is great for configuration. It provides
virtually immediate feedback.
But for larger projects, this can be a double-edged sword, leaving your plugin
susceptible to unexpected bugs at the wrong time.

You can leverage LuaCATS https://luals.github.io/wiki/annotations/
annotations, along with lua-language-server https://luals.github.io/ to catch
potential bugs in your CI before your plugin's users do.

------------------------------------------------------------------------------
Tools						 *lua-plugin-type-safety-tools*

- lua-typecheck-action https://github.com/marketplace/actions/lua-typecheck-action
- llscheck https://github.com/jeffzi/llscheck
- luacheck https://github.com/mpeterv/luacheck for additional linting

> - neodev.nvim https://github.com/folke/neodev.nvim

Reviewer 9: Given that neovim now contains most type annotations is there still a reason for neodev? (Other than the luals configuration bits, which we could point out in the luals reference above)

Author: It can still be useful for getting lua-ls diagnostics for plugin dependencies.

I personally prefer to generate a .luarc.json, but that uses Nix and would probably be too niche/overwhelming to recommend in nvim docs.

==============================================================================

> User commands					     *lua-plugin-user-commands*

Reviewer 9: I agree on the UI aspect, but I think we should first do the homework and come up with a better solution for subcommand definitions instead of spreading a chunk of boilerplate for every plugin to copy and paste. 

My vote goes for leaving this section out for now. But let's see what the others say.

Author: I definitely agree that the UI/boilderplate is horrible for plugin authors.
From a user perspective, I find scoped sub commands a lot easier to handle.

How about we open an issue for a better UI? I'd be more than happy to draft a POC if I can find some time soon.

Reviewer 9: > How about we open an issue for a better solution?

Sounds good to me

Author: https://github.com/neovim/neovim/pull/29116

Reviewer 10: @mrcjkb maybe it more clearly, for multip unreadable commands, we should use sub command. if only two comamnd, `FooInstall` and `FooUninstall`, maybe no need to do this.

BTW, I think if this is meregd, it is not only for lua command, but also viml user command. so maybe the guide need to remove `lua-` prefix

Author: I've opened https://github.com/neovim/neovim/issues/32263 to kickstart the discussion about user commands.
Given that @ColinKennedy [has a reference implementation that could potentially be extracted into a library](https://github.com/neovim/neovim/pull/29073#issuecomment-2439674283), and given that I don't see any movement on that topic happening in Nvim core anytime soon, I second the proposal to leave this section out for now.
I'll leave it in the nvim-neorocks/nvim-best-practices guide, with a reference to the implementation in the plugin template.

            if cmdline:match("^Foo[!]*%s+%w*$") then
                -- Filter subcommands that match
                local subcommand_keys = vim.tbl_keys(subcommand_tbl)
                return vim.iter(subcommand_keys)
                    :filter(function(key)
                        return key:find(arg_lead) ~= nil
                    end)
                    :totable()
            end
        end,
        bang = true, -- If you want to support ! modifiers
    })
<
==============================================================================
Keymaps							   *lua-plugin-keymaps*

Avoid creating keymaps automatically, unless they are not controversial. Doing
so can easily lead to conflicts with user |mapping|s.

NOTE: An example for uncontroversial keymaps are buffer-local |mapping|s for 
      specific file types or floating windows.

A common approach to allow keymap configuration is to define a declarative DSL
https://en.wikipedia.org/wiki/Domain-specific_language via a `setup` function.

However, doing so means that

- You will have to implement and document it yourself.
- Users will likely face inconsistencies if another plugin has a slightly
  different DSL.

> - |init| scripts that call such a `setup` function may throw an error if the 

Reviewer 8: ```suggestion
- |init.lua| scripts that call such a `setup` function may throw an error if the 
```

  plugin is not installed or disabled.

As an alternative, you can provide |<Plug>| mappings to allow users to define
their own keymaps with |vim.keymap.set()|.

- This requires one line of code in user configs.
- Even if your plugin is not installed or disabled, creating the keymap won't
  throw an error.

> Another option is to simply expose a Lua function or a |user-command|.

Reviewer 8: ```suggestion
Another option is to simply expose a Lua function or a |user-commands|.
```

<
In the user's config:
>lua
    vim.keymap.set({"n", "v"}, "<leader>h", "<Plug>(SayHello)")
<
==============================================================================
Initialization					    *lua-plugin-initialization*

Newcomers to Lua plugin development will often put all initialization logic in
a single `setup` function, which takes a table of options.
If you do this, users will be forced to call this function in order to use
your plugin, even if they are happy with the default configuration.

Cleanly separated configuration and smart initialization allow your plugin to 
work out of the box.

NOTE: A well designed plugin has minimal impact on startup time.
      See also |lua-plugin-lazy-loading|.

Common approaches to a cleanly separated configuration are:

- A Lua function, e.g. `setup(opts)` or `configure(opts)`, which only overrides the
  default configuration and does not contain any initialization logic.
- A Vimscript compatible table (e.g. in the |vim.g| or |vim.b| namespace) that your
  plugin reads from and validates at initialization time.
  See also |lua-vim-variables|.

Typically, automatic initialization logic is done in a |plugin| or |ftplugin|

> script. See also |runtimepath|.

Reviewer 8: ```suggestion
script. See also |'runtimepath'|.
```

> Newcomers to Lua plugin development will often put all initialization logic in
> a single `setup` function, which takes a table of options.
> If you do this, users will be forced to call this function in order to use
> your plugin, even if they are happy with the default configuration.
> 
> Cleanly separated configuration and smart initialization allow your plugin to 
> work out of the box.
> 
> NOTE: A well designed plugin has minimal impact on startup time.
>       See also |lua-plugin-lazy-loading|.
> 
> Common approaches to a cleanly separated configuration are:
> 
> - A Lua function, e.g. `setup(opts)` or `configure(opts)`, which only overrides the
>   default configuration and does not contain any initialization logic.
> - A Vimscript compatible table (e.g. in the |vim.g| or |vim.b| namespace) that your
>   plugin reads from and validates at initialization time.
>   See also |lua-vim-variables|.
> 
> Typically, automatic initialization logic is done in a |plugin| or |ftplugin|
> script. See also |'runtimepath'|.

Reviewer 11: Although technically correct (newcomers indeed often do that), I would like to ask to rephrase the "intro" or remove it completely due to implied negative connotations ("newcomers often don't know what they are doing and they do this, so ...").

Same with using "Cleanly separated ..." (implying that "not separated is not clean"). I'd suggest using either "explicitly separated", "intentionally separated", or just "separated" without any description.

Here is my suggestion:

```suggestion
Separated configuration and smart initialization allow your plugin to work out
of the box. Common approaches are:

- A Lua function, e.g. `setup(opts)` or `configure(opts)`, which only overrides
  the default configuration and does not contain any initialization logic.
- A Vimscript compatible table (e.g. in the |vim.g| or |vim.b| namespace) that
  your plugin reads from and validates at initialization time.
  See also |lua-vim-variables|.

Typically, automatic initialization logic is done in a |plugin| or |ftplugin|
script. See also |'runtimepath'|.

If you want to force users to explicitly enable plugin functionality, using
`setup(opts)` function for both initialization and configuration is a common
practice.

NOTE: A well designed plugin has minimal impact on startup time.
      See also |lua-plugin-lazy-loading|.
```

Reviewer 7: >  I would like to ask to rephrase the "intro" or remove it completely due to implied negative connotations

I don't know where you read this in between the lines
I think it's perfectly fine and there is value in describing what the current "de facto standard" is and why that's not good

> Same with using "Cleanly separated ..." (implying that "not separated is not clean"). I'd suggest using either "explicitly separated", "intentionally separated", or just "separated" without any description.

As far as I know as a non-native English speaker "clean separation" doesn't at all imply that otherwise it's not clean (which imo even is the case here but that's not the discussion right now)
as duckduckgo ai chat said:
> The term "clean separation" typically refers to a clear and distinct division or distinction between two things or entities. It does not necessarily imply cleanliness in the literal sense, but rather emphasizes a clear and unambiguous separation.

I found "separated configuration" quite a confusing term though
For me this more sounds like you'd want people to split up the configuration even further

Reviewer 11: > I think it's perfectly fine and there is value in describing what the current "de facto standard" is and why that's not good

Less opinionated -> more better. A suggested rewrite contains both emphasis on configuration-initialization separation and the "de facto standard" which is a perfectly fine alternative if understanding the consequences.

> I found "separated configuration" quite a confusing term though

It is not "<separated configuration> and <initialization>" but more like "configuration and initialization are separate". Both sound OK to me.

Author: Given our previous negative interaction, I will give one response and then leave it to everyone else to decide.

I disagree with your suggested change for the following reasons:

It is important to emphasize the consequences of defaulting to forcing users to call a `setup` function in the way I am doing As @max397574 points out, it's a fact that newcomers usually default to this without understanding the consequences.
As mentioned in https://github.com/nvim-neorocks/nvim-best-practices/issues/5#issuecomment-2131367234, plugin authors can choose to diverge if they deem it necessary, which is made clear right at the beginning of this guide:

> it is intended as a set of rules...

I disagree that the way I have phrased it has any negative connotations (this is a nitpick), as I don't make any recommendations not to diverge.

I second @max397574's statement on "clean separation". But I see how this could be misinterpreted as negative for non-native speakers.
How about "strict separation" instead?

Reviewer 11: > It is important to emphasize the consequences of defaulting to forcing users to call a `setup` function

It *is* described in the suggested change: "If you want to force users to explicitly enable plugin functionality, using ...". Without unnecessary "Newcomers often" / "even if they are happy", as built-in docs usually prefer to be more on point.

> How about "strict separation" instead?

To me this is also fine. Using no description is even better.

Author: I strongly disagree that your suggestion "is more on point".
The sentence as you suggest would be redundant, as it lacks any reason for it, while losing the *emphasis* on the *very often overlooked* consequences. I have often heard users complain about being forced to call a `setup` function. I have not once had anyone complain about *not being forced* to call one. 

If anything, the phrasing, "Newcomers will often..." makes it clear that this advice is directed at newcomers, knowing that more advanced plugin developers will be able to make their own judgement.

> To me this is also fine. 

I'll adjust it later today.

  implementation.

NOTE: If you have a function that takes a large options table, creating lots
      of |<Plug>| mappings to expose all of its uses could become
      overwhelming. It may still be beneficial to create some for the most
      common ones.

------------------------------------------------------------------------------
Example					      *lua-plugin-plug-mapping-example*

In your plugin:
>lua
    vim.keymap.set("n", "<Plug>(SayHello)", function() 
        print("Hello from normal mode") 
    end, { noremap = true })

    vim.keymap.set("v", "<Plug>(SayHello)", function() 
        print("Hello from visual mode") 
    end, { noremap = true })
<
In the user's config:
>lua
    vim.keymap.set({"n", "v"}, "<leader>h", "<Plug>(SayHello)")
<
==============================================================================
Initialization					    *lua-plugin-initialization*

Newcomers to Lua plugin development will often put all initialization logic in
a single `setup` function, which takes a table of options.
If you do this, users will be forced to call this function in order to use
your plugin, even if they are happy with the default configuration.

> Cleanly separated configuration and smart initalization allow your plugin to 

Reviewer 8: ```suggestion
Cleanly separated configuration and smart initialization allow your plugin to 
```

filetype, by putting the initialization logic in a `ftplugin/{filetype}.lua`
script.

------------------------------------------------------------------------------
Example				     *lua-plugin-lazy-loading-filetype-example*

A plugin tailored to Rust development might have initialization in 
`ftplugin/rust.lua`:
>lua
    if not vim.g.loaded_my_rust_plugin then
        -- Initialise
    end
    -- NOTE: Using `vim.g.loaded_` prevents the plugin from initializing twice
    -- and allows users to prevent plugins from loading
    -- (in both Lua and Vimscript).
    vim.g.loaded_my_rust_plugin = true

    local bufnr = vim.api.nvim_get_current_buf()
    -- do something specific to this buffer,
    -- e.g. add a |<Plug>| mapping or create a command
    vim.keymap.set("n", "<Plug>(MyPluginBufferAction)", function() 
        print("Hello")
    end, { noremap = true, buffer = bufnr, })
<
==============================================================================
Configuration				             *lua-plugin-configuration*

Once you have merged the default configuration with the user's config, you
should validate configs.

Validations could include:

> - Correct types, see |vim.validate()|

Reviewer 8: neovim just got a flat version(?) that is much faster than this. Wonder if we should mention that instead/as well since vim.validate has significant performance impact? Unsure, just mentioning for reference. @gpanders might have some input on this.

Author: I suppose you mean this: https://github.com/neovim/neovim/pull/28977

- Unknown fields in the user config (e.g. due to typos).
  This can be tricky to implement, and may be better suited for a |health| 
  check, to reduce overhead.

> NOTE: |vim.validate()| will `error` if it fails valdation.

Reviewer 8: ```suggestion
NOTE: |vim.validate()| will `error` if it fails validation.
```

    ---@return boolean is_valid
    ---@return string|nil error_message
    local function validate_path(path, tbl)
      local ok, err = pcall(vim.validate, tbl)
      return ok, err and path .. "." .. err
    end
>
The function can be called like this:
>lua
    ---@param cfg myplugin.Config
    ---@return boolean is_valid
    ---@return string|nil error_message
    function validate(cfg)
        return validate_path("vim.g.my_plugin", {
            do_something_cool = { cfg.do_something_cool, "boolean" },
            strategy = { cfg.strategy, "string" },
        })
    end
<
An invalid config will result in an error message like 
`"vim.g.my_plugin.strategy: expected string, got number"`.

By doing this, you can use the validation both

- At initialization, reporting errors with |vim.notify()|.
- In a |health| check.

==============================================================================
Troubleshooting					   *lua-plugin-troubleshooting*

------------------------------------------------------------------------------
Health checks			       	    *lua-plugin-troubleshooting-health*

Provide health checks in `lua/{plugin}/health.lua`.

Some things to validate:

> - User cnfiguration

Reviewer 8: ```suggestion
- User configuration
```

- Luarocks <3 Neovim https://github.com/nvim-neorocks/sample-luarocks-plugin

------------------------------------------------------------------------------
Tools			                 *lua-plugin-versioning-releases-tools*

- luarocks-tag-release
  https://github.com/marketplace/actions/luarocks-tag-release
- release-please-action
  https://github.com/marketplace/actions/release-please-action
- semantic-release 
  https://github.com/semantic-release/semantic-release 

==============================================================================
Documentation                                        *lua-plugin-documentation*

Provide vimdoc (see |help-writing|), so that users can read your plugin's
documentation in Neovim, by entering `:h {plugin}` in |command-mode|.

------------------------------------------------------------------------------
Tools			                       *lua-plugin-documentation-tools*

- panvimdoc https://github.com/kdheepak/panvimdoc

------------------------------------------------------------------------------
Further reading			     *lua-plugin-documentation-further-reading*

- Diátaxis - A systematic approach to technical documentation authoring

>   https://diataxis.fr/

Reviewer 8: Probably not bad, but it looks like a snoozefest

script. See also |runtimepath|.

==============================================================================
Lazy loading					      *lua-plugin-lazy-loading*

When it comes to initializing your plugin, assume your users may not be using
a plugin manager that takes care of lazy loading for you.
Making sure your plugin does not unnecessarily impact startup time is your
responsibility. A plugin's functionality may evolve over time, potentially
leading to breakage if users have to hack into the loading mechanisms.
Furthermore, a plugin that implements its own lazy initialization properly will
likely have less overhead than the mechanisms used by a plugin manager or user
to load that plugin lazily.

Think carefully about when which parts of your plugin need to be loaded.

------------------------------------------------------------------------------
Defer `require` calls			*lua-plugin-lazy-loading-defer-require*

|plugin| scripts should not eagerly `require` Lua modules.

For example, instead of:
>lua
    local foo = require("foo")
    vim.api.nvim_create_user_command("MyCommand", function()
        foo.do_something()
    end, {
      -- ...
    })
<

> ...which will eagerly load the `foo` module and any other modules it imports
> eagerly, can lazy load it by moving the `require` into the command's
> implementation.

Reviewer 8: ```suggestion
which will eagerly load the `foo` module and any other modules it imports
eagerly, you can lazy load it by moving the `require` into the command's
implementation.
```

Common approaches to a cleanly separated configuration are:

- A Lua function, e.g. `setup(opts)` or `configure(opts)`, which only overrides the
  default configuration and does not contain any initialization logic.
- A Vimscript compatible table (e.g. in the |vim.g| or |vim.b| namespace) that your
  plugin reads from and validates at initialization time.
  See also |lua-vim-variables|.

Typically, automatic initialization logic is done in a |plugin| or |ftplugin|
script. See also |runtimepath|.

==============================================================================
Lazy loading					      *lua-plugin-lazy-loading*

When it comes to initializing your plugin, assume your users may not be using
a plugin manager that takes care of lazy loading for you.
Making sure your plugin does not unnecessarily impact startup time is your
responsibility. A plugin's functionality may evolve over time, potentially
leading to breakage if users have to hack into the loading mechanisms.
Furthermore, a plugin that implements its own lazy initialization properly will
likely have less overhead than the mechanisms used by a plugin manager or user
to load that plugin lazily.

> Think carefully about when which parts of your plugin need to be loaded.

Reviewer 8: Annoying sentence. "Think smart and lots, don't be dumb"

Author: I guess I could just remove it? The relevant info is in the subsections that follow after it.

Reviewer 8: I think so. I'm aware this is well into nitpick territory but whatever, you'll live

Author: I'm very happy for your feedback :pray:  :smile:

==============================================================================
User commands					     *lua-plugin-user-commands*

Many users rely on command completion to discover available user commands. If
a plugin pollutes the command namespace with lots of commands, this can
quickly become overwhelming.

Example:

- `FooAction1 {arg}`
- `FooAction2 {arg}`
- `FooAction3`
- `BarAction1`
- `BarAction2`

Instead of doing this, consider gathering subcommands under scoped commands
and implementing completions for each subcommand.

Example:

- `Foo action1 {arg}`
- `Foo action2 {arg}`
- `Foo action3`
- `Bar action1`
- `Bar action2`

------------------------------------------------------------------------------
Subcommand completions example   *lua-plugin-user-commands-completions-example*

In this example, we want to provide:

- Subcommand completions if the user has typed `:Foo ...`
- Argument completions if they have typed `:Foo {subcommand}`

First, define a type for each subcommand, which has:

- An implementation: A function which is called when executing the subcommand.
- An optional command completion callback, which takes the lead of the
  subcommand's arguments.

>lua

> ---@class FooSubcommand
> ---@field impl fun(args:string[], opts: table)
> ---@field complete? fun(subcmd_arg_lead: string): string[]

Reviewer 8: You need to indent all codeblocks for TS to highlight correctly. Don't ask me why, thems the rules.

```suggestion
    ---@class FooSubcommand
    ---@field impl fun(args:string[], opts: table)
    ---@field complete? fun(subcmd_arg_lead: string): string[]
```

Author: Ah, I was wondering about that :smile: 
Good to know.

Further reading		       *lua-plugin-versioning-releases-further-reading*

- Luarocks <3 Neovim https://github.com/nvim-neorocks/sample-luarocks-plugin

------------------------------------------------------------------------------
Tools			                 *lua-plugin-versioning-releases-tools*

- luarocks-tag-release
  https://github.com/marketplace/actions/luarocks-tag-release
- release-please-action
  https://github.com/marketplace/actions/release-please-action
- semantic-release 
  https://github.com/semantic-release/semantic-release 

==============================================================================
Documentation                                        *lua-plugin-documentation*

Provide vimdoc (see |help-writing|), so that users can read your plugin's
documentation in Neovim, by entering `:h {plugin}` in |command-mode|.

------------------------------------------------------------------------------
Tools			                       *lua-plugin-documentation-tools*

- panvimdoc https://github.com/kdheepak/panvimdoc
- lemmy-help https://github.com/numToStr/lemmy-help

NOTE: lemmy-help has not been updated recently and is no longer fully

>       compatible with |LuaCATS|.

Author: Not sure if we want to add this here.
It hasn't been touched in about 2 years. I still use it and it works fine (as long as I don't use newer features like `(exact)` classes, but I'm either going to fork it or switch to something else.

Reviewer 8: LMAO absolutely remove this what

Reviewer 8: If it's deprecated and not recommended then people don't need to know. Just yeet it.

*lua-plugin.txt*                     Nvim

                            NVIM REFERENCE MANUAL

		  Guide to developing Lua plugins for Nvim

                                       Type |gO| to see the table of contents.

==============================================================================
Introduction                                                       *lua-plugin*

This is a guide for getting started with Neovim plugin development. It is not
intended as a set of rules, but as a collection of recommendations for good
practices.

For a guide to using Lua in Neovim, please refer to |lua-guide|.

==============================================================================
Type safety                                            *lua-plugin-type-safety*

Lua, as a dynamically typed language, is great for configuration. It provides
virtually immediate feedback.
But for larger projects, this can be a double-edged sword, leaving your plugin
susceptible to unexpected bugs at the wrong time.

You can leverage |LuaCATS| https://luals.github.io/wiki/annotations/
annotations, along with lua-language-server https://luals.github.io/ to catch

> potential bugs in your CI before your plugin's users do

Reviewer 8: ```suggestion
potential bugs in your CI before your plugin's users do.
```

------------------------------------------------------------------------------
Tools						 *lua-plugin-type-safety-tools*

- lua-typecheck-action https://github.com/marketplace/actions/lua-typecheck-action
- llscheck https://github.com/jeffzi/llscheck
- luacheck https://github.com/mpeterv/luacheck for additional linting
- neodev.nvim https://github.com/folke/neodev.nvim

------------------------------------------------------------------------------
Further reading			       *lua-plugin-type-safety-further-reading*

> - LuaCATS documentation https://luals.github.io/wiki/annotations/     *LuaCATS*

Reviewer 8: You already linked to this before. This doesn't seem necessary.
